Dockerfile

```diff
@@ -1,56 +1,129 @@
+RUN git clone --depth 1 --branch llvmorg-13.0.0 https://github.com/llvm/llvm-project.git
+RUN git clone --depth 1 --branch master https://github.com/trailofbits/blight.git
+
+RUN pip3 install pytest /blight
 
 RUN update-ca-certificates
 
-RUN apt-get install -y software-properties-common && \
-      # deadsnakes PPA allows us to install python3.10 on focal
-      add-apt-repository ppa:deadsnakes/ppa && \
-      apt-get -y update && \
-      apt-get install -y \
-            python3.10-dev \
-            # make update-alternatives work
-            python3.10-distutils \
-            curl && \
-      # need a pip compatible with 3.10
-      curl -sS https://bootstrap.pypa.io/get-pip.py | python3.10 && \
-      update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 10 && \
-      python3 -m pip install pytest
-
-WORKDIR /blight
-RUN git clone https://github.com/trailofbits/blight.git .
-RUN pip3 install .
+RUN GO111MODULE=off go get github.com/SRI-CSL/gllvm/cmd/...
+ENV PATH="$PATH:/root/go/bin"
+
+FROM base as clean-libcxx
+
+ENV BITCODE=/cxx_clean_bitcode
+RUN mkdir -p $BITCODE
+ENV WLLVM_BC_STORE=$BITCODE
+
+ENV CC="gclang"
+ENV CXX="gclang++"
+
+ENV LIBCXX_BUILD_DIR=/llvm-project/build
+ENV LIBCXX_INSTALL_DIR=/cxx_lib/clean_build
+
+RUN cmake -GNinja \
+  -B$LIBCXX_BUILD_DIR \
+  -S/llvm-project/runtimes \
+  -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+  -DCMAKE_INSTALL_PREFIX=$LIBCXX_INSTALL_DIR \
+  -DLIBCXXABI_ENABLE_SHARED=NO \
+  -DLIBCXX_ENABLE_SHARED=NO \
+  -DLIBCXX_CXX_ABI="libcxxabi" \
+  -DLLVM_ENABLE_LIBCXX=ON \
+  -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi"
+
+RUN cmake --build $LIBCXX_BUILD_DIR --target install-cxx install-cxxabi -j$((`nproc`+1))
+
+FROM clean-libcxx as polytracker-libcxx
 
+ENV BITCODE=/cxx_poly_bitcode
+RUN mkdir -p $BITCODE
+ENV WLLVM_BC_STORE=$BITCODE
+
+ENV CC="gclang"
+ENV CXX="gclang++"
+
+ENV LIBCXX_BUILD_DIR=/llvm-project/llvm/build
+ENV LIBCXX_INSTALL_DIR=/cxx_lib/poly_build
+
+RUN cmake -GNinja \
+  -B$LIBCXX_BUILD_DIR \
+  -S/llvm-project/runtimes \
+  -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+  -DCMAKE_INSTALL_PREFIX=$LIBCXX_INSTALL_DIR \
+  -DLIBCXX_ABI_NAMESPACE="__p" \
+  -DLIBCXXABI_ENABLE_SHARED=NO \
+  -DLIBCXX_ENABLE_SHARED=NO \
+  -DLIBCXX_ABI_VERSION=2 \
+  -DLIBCXX_CXX_ABI="libcxxabi" \
+  -DLIBCXX_HERMETIC_STATIC_LIBRARY=ON \
+  -DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON \
+  -DLLVM_ENABLE_LIBCXX=ON \
+  -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi"
+
+RUN cmake --build $LIBCXX_BUILD_DIR --target install-cxx install-cxxabi -j$((`nproc`+1))
+
+FROM polytracker-libcxx as polytracker-python
+
+WORKDIR /workdir
 COPY . /polytracker
 
-RUN mkdir /polytracker/build
-WORKDIR /polytracker/build
-RUN cmake -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_VERBOSE_MAKEFILE=TRUE -DCXX_LIB_PATH=/cxx_libs ..
-RUN ninja install
+RUN pip3 install /polytracker
 
+FROM polytracker-python as polytracker-cxx
 
-WORKDIR /polytracker
-RUN pip3 install .
+ARG DFSAN_FILENAME_ARCH=x86_64
```

Comment:
Are we able to build this on other archs? We used to have some (limited) support for arm-builds, I think.

Reply:
I don't see a reason why we wouldn't be able to. That being said, I fully expect there to be issues.

Reply:
Thinking about this a little bit more. Do we have use-cases where we want to build for say ARM on an X86 host? Cause if we omit the target argument altogether it will by default build for the host arch.

Reply:
Iirc, @josephsweeney would like to build on Apple Silicon instead of spinning up a cloud Linux VM or similar. But I don't think that should require ARM builds from an x86 host!

Reply:
Yeah, I don't expect any cross-architecture build necessary. Just making it possible to build on both x86 and arm64 would be good. I also don't think this has to happen immediately, we can get an arm build working when needed.